contrib/xor-blocks/src/main.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+use std::{
+    env,
+    fmt::{self, Debug, Display, Formatter},
+    fs::{self, File},
+    hash::{BuildHasher, Hasher, RandomState},
+    io::{self, BufReader, BufWriter, Read, Write},
+    path::Path,
+    time::{Duration, Instant},
+};
+
+const MAGIC: [u8; 4] = [0xf9, 0xbe, 0xb4, 0xd9];
+const XOR_FILE_NAME: &str = "xor.dat";
+
+#[derive(Debug)]
+struct Error {
+    msg: String,
+}
+
+impl Error {
+    fn new(msg: String) -> Self {
+        Self { msg }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.msg)
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(value: io::Error) -> Self {
+        Error::new(value.to_string())
+    }
+}
+
+impl From<&str> for Error {
+    fn from(value: &str) -> Self {
+        Error::new(value.to_string())
+    }
+}
+
+fn main() -> Result<(), Error> {
+    let start = Instant::now();
+
+    let data_dir_path = if env::args().len() == 2 {
+        let arg = env::args().nth(1).expect("the arg to exist");
+        if !arg.starts_with("--datadir=") && !arg.starts_with("-datadir=") {
+            return Err(Error::new(format!(
+                "Error parsing command line arguments: Invalid parameter {arg}"
+            )));
+        }
+        let (_, datadir) = arg.split_once("-datadir=").expect("split to be some");
+        datadir.into()
+    } else if env::args().len() == 1 {
+        #[allow(deprecated)]
+        env::home_dir()
+            .expect("to have a home directory")
+            .join(match env::consts::OS {
+                "macos" => "Library/Application Support/Bitcoin",
+                "windows" => "AppData\\Roaming\\Bitcoin",
+                "linux" => ".bitcoin",
+                _ => {
+                    return Err("Unknown OS".into());
+                }
+            })
+    } else {
+        return Err("Invalid number of arguments. Run with `cargo run --release` if using default datadir. Otherwise run with `cargo run --release -- -datadir=<custom datadir>`.".into());
+    };
+
+    let blocks_path = data_dir_path.join("blocks");
```

Comment:
Reminds me that non-mainnet use paths like `/regtest/blocks`, perhaps we should accept `-chain` argument as well.